src/contract_client.rs

```diff
@@ -668,18 +671,253 @@ impl ModuleDeployHandle {
+    CallContractMissingContract = -2147483631,
+    #[error("[CallContractError::MissingEntrypoint]")]
+    CallContractMissingEntrypoint = -2147483630,
+    #[error("[CallContractError::MessageFailed]")]
+    CallContractMessageFailed = -2147483629,
+    #[error("[CallContractError::LogicReject]")]
+    CallContractLogicReject = -2147483628,
+    #[error("[CallContractError::Trap]")]
+    CallContractTrap     = -2147483627,
+    #[error("[UpgradeError::MissingModule]")]
+    UpgradeMissingModule = -2147483626,
+    #[error("[UpgradeError::MissingContract]")]
+    UpgradeMissingContract = -2147483625,
+    #[error("[UpgradeError::UnsupportedModuleVersion]")]
+    UpgradeUnsupportedModuleVersion = -2147483624,
+    #[error("[QueryAccountBalanceError]")]
+    QueryAccountBalanceError = -2147483623,
+    #[error("[QueryContractBalanceError]")]
+    QueryContractBalanceError = -2147483622,
+}
+
+/// Decode the `reject_reason` into a human-readable error based on the error
+/// code definition in the `concordium-std` crate.
+pub fn decode_concordium_std_error(reject_reason: i32) -> Option<ConcordiumStdRejectReason> {
+    if (-2147483648..=-2147483622).contains(&reject_reason) {
+        let reason: ConcordiumStdRejectReason = unsafe { ::std::mem::transmute(reject_reason) };
+        Some(reason)
+    } else {
+        None
+    }
+}
+
+/// Decode the smart contract logical revert reason and return a human-readable
+/// error string.
+///
+/// If the error is NOT caused by a smart contract logical revert, the
+/// `reject_reason` is already a human-readable error. No further decoding of
+/// the error is needed and as such this function returns `None`.
+/// An example of such a failure (rejected transaction) is the case when the
+/// transaction runs out of energy which is represented by the human-readable
+/// error variant "OutOfEnergy".
+///
+/// Step 1: If the error matches a smart contract logical revert code coming
+/// from the `concordium-std` crate, this function decodes the error based on
+/// the error code definition in the `concordium-std` crate.
+///
+/// Step 2: If the error is caused by a smart contract logical revert coming
+/// from the smart contract itself, this function uses the provided
+/// `error_schema` and `return_value` to decode the `reject_reason` into a
+/// human-readable error string.
+///
+/// Disclaimer: A smart contract can have logic to overwrite/change the meaning
+/// of the error codes as defined in the `concordium-std` crate. While it is not
+/// advised to overwrite these error codes and is rather unusual to do so, this
+/// function decodes the error codes based on the
+/// definitions in the `concordium-std` crate (assuming they have not been
+/// overwritten with other meanings in the smart contract logic). No guarantee
+/// are given as such that the meaning of the decoded reject reason haven't been
+/// altered by the smart contract logic.
+pub fn decode_smart_contract_revert(
```

Comment:
Would it make sense to move this entire function to `impl RejectReason` or would that seem out of place? It would probably make it easier to discover for users